src/vs/workbench/api/node/extHostTreeExplorers.ts

```diff
@@ -0,0 +1,101 @@
+
+	constructor(
+		threadService: IThreadService,
+		private commands: ExtHostCommands
+	) {
+		super();
+
+		this._proxy = threadService.get(MainContext.MainThreadExplorers);
+
+		this._treeExplorerNodeProviders = Object.create(null);
+		this._externalNodeMaps = Object.create(null);
+	}
+
+	registerTreeExplorerNodeProvider(providerId: string, provider: TreeExplorerNodeProvider<any>): Disposable {
+		this._proxy.$registerTreeExplorerNodeProvider(providerId);
+		this._treeExplorerNodeProviders[providerId] = provider;
+
+		return new Disposable(() => {
+			delete this._treeExplorerNodeProviders[providerId];
+		});
+	}
+
+	$provideRootNode(providerId: string): TPromise<InternalTreeExplorerNode> {
+		const provider = this.getProvider(providerId);
+
+		return asWinJsPromise(() => provider.provideRootNode()).then(externalRootNode => {
+			const treeNodeMap = Object.create(null);
+			this._externalNodeMaps[providerId] = treeNodeMap;
+
+			const internalRootNode = new InternalTreeExplorerNode(externalRootNode, provider);
+			this._externalNodeMaps[providerId][internalRootNode.id] = externalRootNode;
+			return internalRootNode;
+		}, err => {
+			this.showErrorMessage(`TreeExplorerNodeProvider '${providerId}' failed to provide root node.`);
+		});
+	}
+
+	$resolveChildren(providerId: string, mainThreadNode: InternalTreeExplorerNode): TPromise<InternalTreeExplorerNode[]> {
+		const provider = this.getProvider(providerId);
+
+		const externalNodeMap = this._externalNodeMaps[providerId];
+		const externalNode = externalNodeMap[mainThreadNode.id];
+
+		return asWinJsPromise(() => provider.resolveChildren(externalNode)).then(children => {
+			return children.map(externalChild => {
+				const internalChild = new InternalTreeExplorerNode(externalChild, provider);
+				externalNodeMap[internalChild.id] = externalChild;
+				return internalChild;
+			});
+		}, err => {
+			this.showErrorMessage(`TreeExplorerNodeProvider '${providerId}' failed to resolve children.`);
+		});
+	}
+
+	$executeCommand(providerId: string, mainThreadNode: InternalTreeExplorerNode): TPromise<void> {
+		if (mainThreadNode.clickCommand) {
+			const externalNode = this._externalNodeMaps[providerId][mainThreadNode.id];
+			return asWinJsPromise(() => this.commands.executeCommand(mainThreadNode.clickCommand, externalNode)).then(() => {
+				return null;
+			}, err => {
+				this.showErrorMessage(`Failed to execute command '${mainThreadNode.clickCommand}' provided by TreeExplorerNodeProvider '${providerId}'.`);
+			});
+		}
+
+		return TPromise.as(null);
+	}
+
+	getProvider(providerId: string): TreeExplorerNodeProvider<any> {
+		const provider = this._treeExplorerNodeProviders[providerId];
+		if (!provider) {
+			this.showErrorMessage(`No TreeExplorerNodeProvider with id '${providerId}' registered.`);
+		}
+
+		return provider;
+	}
+
+	private showErrorMessage(message: string): void {
+		this._proxy.$showMessage(Severity.Error, message);
```

Comment:
You could probably make ExtHostMessageService a dependency and call into that from here?